File: server/core/src/dtos/Page.kt
package com.lorenzoog.projectkob.core.dtos

import kotlinx.serialization.Serializable

@Serializable
data class Page<T>(
  val items: Collection<T>,
  val totalPages: Int,
)

fun <T> Collection<T>.asPage(pages: Number): Page<T> {
  return Page(this, pages.toInt())
}

inline fun <T, R> Page<T>.map(func: (T) -> R): Page<R> {
  return Page(items.map(func), totalPages)
}

File: web/src/entities/page.ts
class Page<T> {
  public constructor(
    public readonly items: T[],
    public readonly totalPages: number
  ) {}

  public map<R>(fmap: (item: T) => R): Page<R> {
    return new Page(this.items.map(fmap), this.totalPages);
  }

  public filter(filter: (item: T) => boolean): Page<T> {
    return new Page(this.items.filter(filter), this.totalPages);
  }

  public append(item: T): Page<T> {
    return new Page([...this.items, item], this.totalPages);
  }
}

export default Page;
